### PR TITLE
Remove PathSearchTrimEnd

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

This is moved from #18154.

PathSearchTrimEnd is removed based on https://github.com/dotnet/runtime/issues/76122#issuecomment-1256977256
For ex., now we don't remove trail spaces that are valid in file names on Windows.
On Linux it was even more incorrect.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 1]
```diff
@@ -1473,7 +1473,7 @@ private static CanDoPathLookupResult CanDoPathLookup(string possiblePath)
         /// <summary>
         /// The command name to search for.
         /// </summary>
-        private string _commandName;
+        private readonly string _commandName;
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/CommandSearcher.cs:1476]: For reviewers: this fixes a violation `IDE0044`: Add readonly modifier, enabled in #13880.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 2]
```diff
@@ -1561,7 +1560,6 @@ private void setupPathSearcher()
 
                     if (!string.IsNullOrEmpty(fileName))
                     {
-                        fileName = fileName.TrimEnd(Utils.Separators.PathSearchTrimEnd);
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandSearcher.cs:1564]: Test what? In common, it is impossible to know what is OS and file system limitations. As .Net team concluded, the best thing we can do is to defer checks to OS and file system.

> With this we should now be able to call executables that have trailing whitespace? (although seems like a bad practice)

I don't see any changes in behavior on Windows in my manual tests.
I'd expect that now we can run a file with trailing spaces.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
This is a breaking change: today, I can run `& 'pwsh  '` (trailing space in the command name). With this change, it won't work anymore.

I don't think we should change anything in the command searching code. If our `FileSystemProvider` cannot find a file with trailing spaces in its name as it should, then make the changes there, but not the command discovery code.

**Comment by iSazonov** [comment]
> This is a breaking change: today, I can run `& 'pwsh '` (trailing space in the command name). With this change, it won't work anymore.
> 
> I don't think we should change anything in the command searching code. If our `FileSystemProvider` cannot find a file with trailing spaces in its name as it should, then make the changes there, but not the command discovery code.

This is as strange as if we cut the end `H` in `pwsh` and still `pws` would run as `pwsh`.

This rather indicates that there is a bug in the way we run applications and not in the search or FileSystemProvider.

**Pull request closed without merging**